### PR TITLE
Arachne 1293 selenium for study edit

### PR DESCRIPTION
**Pull request by mpozhidaeva**
first part of tests

### CODE REVIEW

File: test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java [hunk 1]
```diff
@@ -1,61 +1,699 @@
...
+    }
+
+    @Test
+    public void test16UpdateStudyStatus() throws Exception {
+
+        loginAndOpenStudy(STUDY_NAME);
+        String studyStatus = "Active";
+
+        WebElement studyStatusSelect = driver.findElement(
+                ByBuilder.byClassAndText("ac-study-status-option__label", studyStatus));
```
[the remark this answers is not in the document]
Comment (mpozhidaeva) [reply test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java]: I guess this selector is necessary only in this class - do you agree?

File: test/java/com/odysseusinc/arachne/portal/front/utils/ByBuilder.java [hunk 2]
```diff
@@ -70,6 +70,8 @@ public static By toolbar(String title) {
         final String selector = String.format(".//*[(@class='ac-toolbar' " +
                         "or starts-with(@class, 'ac-toolbar ')" +
                         "or contains(@class, ' ac-toolbar ') " +
+                        "or contains(@class, 'ac-study-toolbar-title__title') " +
+                        "or contains(@class, 'ac-toolbar__header') " +
```
[the remark this answers is not in the document]
Comment (mpozhidaeva) [reply test/java/com/odysseusinc/arachne/portal/front/utils/ByBuilder.java]: to use toolbar() to check study toolbar

File: test/java/com/odysseusinc/arachne/portal/front/ProfileManagerTest.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+import org.junit.Test;
+import org.junit.runners.MethodSorters;
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebElement;
+
+@FixMethodOrder(MethodSorters.NAME_ASCENDING)
+public class ProfileManagerTest extends BaseTest {
+
+    @AfterClass
+    public static void afterTest() throws IOException {
```
[the remark this answers is not in the document]
Comment (mpozhidaeva) [reply test/java/com/odysseusinc/arachne/portal/front/ProfileManagerTest.java]: done

File: test/java/com/odysseusinc/arachne/portal/front/ProfileManagerTest.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        WebElement middleNameInput = modal.findElement(ByBuilder.input("Middle Name"));
+        WebElement lastNameInput = modal.findElement(ByBuilder.input("Last Name*"));
+        WebElement saveBtn = modal.findElement(ByBuilder.button("Save"));
+
+        firstNameInput.sendKeys(postfix);
+        middleNameInput.sendKeys(postfix);
+        lastNameInput.sendKeys(postfix);
+
+        saveBtn.click();
+        waitFor(driver, ByBuilder.byClassAndText("ac-profile-view__user-name", StringUtils.repeat("admin" + postfix, " ", 3)));
```
[the remark this answers is not in the document]
Comment (mpozhidaeva) [reply test/java/com/odysseusinc/arachne/portal/front/ProfileManagerTest.java]: do you prefer "admin" + postfix + " " +"admin" + postfix + "admin" + postfix ?

File: test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java [hunk 5]
```diff
@@ -1,61 +1,699 @@
...
+        openParticipantsTab();
+        WebElement remove = driver.findElement(By.className("ac-study-participants-item__action-ico--remove"));
+        remove.click();
+        acceptAlert();
+        waitFor(driver, By.className("ac-study-participants-list"));
+        List<WebElement> participantRows = driver.findElements(By.className("ac-study-participants-list"));
+
+        WebElement participantRow = participantRows.stream()
+                .filter(row -> row.findElement(
+                        ByBuilder.byClassAndText("ac-link", "admin2 admin2 admin2")).isDisplayed())
```
[the remark this answers is not in the document]
Comment (mpozhidaeva) [reply test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java]: done

File: test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java [hunk 6]
```diff
@@ -1,61 +1,699 @@
...
+        driver.findElement(By.className("ac-modal--active")).findElement(By.className("ac-modal__close-ico")).click();
 
-        WebDriverWait waitForCreate = new WebDriverWait(driver, 3);
-        waitForCreate.until((Predicate<WebDriver>) driver -> {
-            return driver.findElement(ByBuilder.toolbar())
-                    .findElement(By.xpath(".//*[contains(@class, 'ac-study-toolbar-title__title') and text()='" + studyName + "']"))
+        webDriverWait = new WebDriverWait(driver, 3);
+        webDriverWait.until((Predicate<WebDriver>) driver -> {
+            return driver
+                    .findElements(By.xpath(".//*[contains(@class, 'ac-select-control__label') and text()='" + updatedRole + "']")).get(1)
```
[the remark this answers is not in the document]
Comment (mpozhidaeva) [reply test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java]: done

File: test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java [hunk 7]
```diff
@@ -1,61 +1,699 @@
...
+        // redirect to analysis page
+        waitFor(driver, ByBuilder.toolbar(analysisName));
+
+        // return on study page -> back button
+        WebElement backToStudyBtn = driver.findElement(By.className("ac-toolbar__back-icon"));
+        backToStudyBtn.click();
+
+        waitFor(driver, ByBuilder.toolbar(STUDY_NAME));
+        Assert.assertTrue(driver.findElement(By.className("ac-study-analyses-list"))
+                .findElement(ByBuilder.byClassAndText("ac-link", analysisName)).isDisplayed());
```
[the remark this answers is not in the document]
Comment (mpozhidaeva) [reply test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java]: done

File: test/java/com/odysseusinc/arachne/portal/front/ProfileManagerTest.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        waitFor(driver, profileMenu);
+        WebElement profileMenuWebElement = driver.findElement(profileMenu);
+        profileMenuWebElement.click();
+
+        By userProfile = ByBuilder.text("User Profile");
+        waitFor(driver, userProfile);
+        WebElement userProfileWebElement = driver.findElement(userProfile);
+        userProfileWebElement.click();
+
+        waitFor(driver, ByBuilder.byClassAndText("ac-profile-view__user-name", "admin admin admin"));
```
Comment (pavgra) [inline test/java/com/odysseusinc/arachne/portal/front/ProfileManagerTest.java]: Still believe that there is better way to find text inside toobar, than stick to concrete and very specific classes

File: test/java/com/odysseusinc/arachne/portal/front/ProfileManagerTest.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        driver.findElement(ByBuilder.byClassAndText("ac-button ac-profile-view__edit-button", "mode_edit")).click();
```
Comment (pavgra) [inline test/java/com/odysseusinc/arachne/portal/front/ProfileManagerTest.java]: Why not iconBtn?

File: test/java/com/odysseusinc/arachne/portal/front/utils/ByBuilder.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (pavgra) [reply test/java/com/odysseusinc/arachne/portal/front/utils/ByBuilder.java]: not the best idea

File: test/java/com/odysseusinc/arachne/portal/front/utils/ByBuilder.java [hunk 10]
```diff
@@ -111,4 +111,30 @@ public static By tableWithHeader(String header) {
...
+
+
+    public static By toolbarEditIco(String title) {
+
+        String selector = ".//*[contains(@class, 'ac-toolbar__edit-ico')" + (title != null ? "and text()='" + title + "'" : "") + "]";
+        return By.xpath(selector);
+    }
+
+
+    public static By tableRow(String placeholder) {
```
[the remark this answers is not in the document]
Comment (pavgra) [reply test/java/com/odysseusinc/arachne/portal/front/utils/ByBuilder.java]: `universal method`
agree

File: test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java [hunk 11]
```diff
@@ -1,61 +1,699 @@
...
-        waitForPageLoad(driver, addBtn);
+        createStudy(STUDY_DATA);
+    }
+
+    @Test
+    public void test02UpdateStudyTitle() throws Exception {
+
+        loginAndOpenStudy(BEFORE_UPDATING_STUDY_NAME);
+
+        By editBtn = ByBuilder.toolbarEditIco("edit");
```
Comment (pavgra) [inline test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java]: Why you name method "toolbarEditIco" and then, again, explicitely pass "edit" param?

File: test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java [hunk 12]
```diff
@@ -1,61 +1,699 @@
...
+        studySaveBtn.click();
+        waitFor(driver, ByBuilder.toolbar(STUDY_NAME));
+    }
+
+    @Test
+    public void test03UpdateStudyDescription() throws Exception {
+
+        loginAndOpenStudy();
+
+        By editBtn = ByBuilder.byClassAndText("ac-material-icons", "edit");
```
Comment (pavgra) [inline test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java]: why not icon/button builder?

File: test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java [hunk 13]
```diff
@@ -1,61 +1,699 @@
...
-        WebElement studyTypeSelect = modalWebElement.findElement(ByBuilder.select(PLACEHOLDER_STUDY_TYPE));
-        WebElement studyTypeOption = modalWebElement.findElement(ByBuilder.selectOption(studyType));
-        WebElement studyCreateBtn = modalWebElement.findElement(ByBuilder.button("Create"));
+        String newDescription = "Long and beautiful description";
+        description.sendKeys(newDescription);
 
-        studyNameInput.sendKeys(studyName);
+        WebElement studyDescriptionBtn = driver
+                .findElement(By.className("ac-study-objective-edit"))
+                .findElement(ByBuilder.byClassAndText("ac-form__submit", "Save"));
```
Comment (pavgra) [inline test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java]: Why not button builder?

File: test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java [hunk 14]
```diff
@@ -1,61 +1,699 @@
...
+
+        String updatedStudyType = "Health Economics and Outcomes";
+        WebElement studyTypeOption = driver.findElement(ByBuilder.selectOption(updatedStudyType));
 
         Actions actions = new Actions(driver);
         actions.moveToElement(studyTypeSelect).click().build().perform();
         actions.moveToElement(studyTypeOption).click().build().perform();
 
-        studyCreateBtn.click();
+        waitFor(driver, ByBuilder.select(updatedStudyType));
```
Comment (pavgra) [inline test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java]: Let's choose single naming: in study title edit you used BEFORE_TITLE and TITLE, here you are using TYPE and UPDATED_TYPE. (I would prefer second convention)

File: test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java [hunk 15]
```diff
@@ -1,61 +1,699 @@
...
+    }
+
+    @Test
+    public void test05AddStudyDocument() throws Exception {
+
+        loginAndOpenStudy();
+
+        WebElement studyTypeSelect = driver.findElement(ByBuilder.byClassAndText("ac-study-document-list-add__label", "Add document"));
+        studyTypeSelect.click();
+        waitFor(driver, ByBuilder.byClassAndText("ac-tabs__item ac-tabs__item--selected", "Computer"));
```
Comment (pavgra) [inline test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java]: Search inside modal

File: test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java [hunk 16]
```diff
@@ -1,61 +1,699 @@
...
+        waitFor(driver, ByBuilder.select(updatedStudyType));
+    }
+
+    @Test
+    public void test05AddStudyDocument() throws Exception {
+
+        loginAndOpenStudy();
+
+        WebElement studyTypeSelect = driver.findElement(ByBuilder.byClassAndText("ac-study-document-list-add__label", "Add document"));
+        studyTypeSelect.click();
```
Comment (pavgra) [inline test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java]: Why is it here?

File: test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java [hunk 17]
```diff
@@ -1,61 +1,699 @@
...
+        waitFor(driver, ByBuilder.byClassAndText("ac-tabs__item ac-tabs__item--selected", "Computer"));
+
+        WebElement uploadElement = driver.findElement(By.cssSelector("input[type=file]"));
+        uploadElement.sendKeys(file.getAbsolutePath());
+        driver.findElement(ByBuilder.byClassAndText("ac-form__submit", "Add")).click();
```
Comment (pavgra) [inline test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java]: Why not button builder?

File: test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java [hunk 18]
```diff
@@ -1,61 +1,699 @@
...
+        waitFor(driver, ByBuilder.byClassAndText("ac-link ac-code-file-info__name", "text.txt"));
+    }
+
+    @Test
+    public void test06RemoveStudyDocument() throws Exception {
+
+        loginAndOpenStudy();
+        waitFor(driver, By.className("ac-list-item__remove"));
```
Comment (pavgra) [inline test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java]: This is too general search. Such remove class can appear in Analysis list, too. SSo find parent panel first and then search inside it

File: test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java [hunk 19]
```diff
@@ -1,61 +1,699 @@
...
+        WebElement removeIco = driver.findElement(By.className("ac-list-item__remove"));
+        removeIco.click();
+        acceptAlert();
+        waitFor(driver, ByBuilder.byClassAndText("ac-list-item__content", "No documents available"));
+    }
+
+    @Test
+    public void test07AddStudyParticipant() throws Exception {
+
+        String chosenName = "admin2 admin2";
```
Comment (pavgra) [inline test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java]: Again, use relative search

File: test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java [hunk 20]
```diff
@@ -1,61 +1,699 @@
...
+        WebElement content = driver.findElement(By.className("ac-notify-dropdown__content"));
+        waitFor(driver, By.className("ac-activity-list-item__author")); //
+        WebElement author = content.findElement(By.className("ac-activity-list-item__author"));
+
+        Assert.assertEquals("admin4 admin4", author.getText());
+        Assert.assertEquals(PORTAL_BASE_URL + "/expert-finder/profile/4", author.getAttribute("href"));
+
+        WebElement study = content.findElement(By.className("ac-activity-list-item__entity"));
+        Assert.assertEquals(STUDY_NAME, study.getText());
+        WebElement accept = content.findElement(ByBuilder.byClassAndText("ac-activity-list-item__action", "Accept"));
```
Comment (pavgra) [inline test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java]: Why not btn builder?

File: test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java [hunk 21]
```diff
@@ -1,61 +1,699 @@
...
+        List<WebElement> addedRows = driver.findElements(By.className("ac-list-item"));
+
+        WebElement participantRoleSelect = addedRows.stream()
+                .filter(row -> row.findElement(
+                        ByBuilder.byClassAndText("ac-link", "admin2 admin2 admin2")).isDisplayed())
+                .findAny()
+                .get();
+
+        String updatedRole = "Contributor";
+        WebElement roleTypeOption = participantRoleSelect.findElement(By.className("ac-select__options"))
```
Comment (pavgra) [inline test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java]: Alex R created builder for selects

File: test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java (hunk 11 above)
Comment (mpozhidaeva) [reply test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java]: "EditIco" is from class name, but text may be different in common case

File: test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java (hunk 14 above)
Comment (mpozhidaeva) [reply test/java/com/odysseusinc/arachne/portal/front/StudyManagerTest.java]: there is only 1 case with BEFORE_TITLE and many with TITLE, so it's seems UPDATED_TITLE in this cases may be not understandable

**Pull request merged**